barretenberg/cpp/src/barretenberg/vm2/simulation/context.hpp

```diff
@@ -66,26 +76,94 @@ class Context : public ContextInterface {
     bool halted() const override { return has_halted; }
     void halt() override { has_halted = true; }
 
+    uint32_t get_context_id() const override { return context_id; }
+
     // Environment.
     const AztecAddress& get_address() const override { return address; }
     const AztecAddress& get_msg_sender() const override { return msg_sender; }
     std::span<const FF> get_calldata() const override { return calldata; }
     bool get_is_static() const override { return is_static; }
 
+    // Event Emitting
+    void emit_context_snapshot() override
+    {
+        ctx_stack_events.emit({ .id = context_id,
+                                .next_pc = next_pc,
+                                .msg_sender = msg_sender,
+                                .contract_addr = address,
+                                .is_static = is_static });
+    };
+
+    ContextEvent get_current_context() override
+    {
+        return {
+            .id = context_id, .pc = pc, .msg_sender = msg_sender, .contract_addr = address, .is_static = is_static
+        };
+    };
+
   private:
     // Environment.
     AztecAddress address;
     AztecAddress msg_sender;
     std::vector<FF> calldata;
     bool is_static;
 
+    uint32_t context_id;
+
     // Machine state.
     uint32_t pc = 0;
     uint32_t next_pc = 0;
     bool has_halted = false;
     std::vector<FF> nested_returndata;
     std::unique_ptr<BytecodeManagerInterface> bytecode;
     std::unique_ptr<MemoryInterface> memory;
+
+    // Emiiters
+    EventEmitterInterface<ContextStackEvent>& ctx_stack_events;
+};
+
+// TODO(ilyas): flesh these out in the cpp file, these are just temporary
+class EnqueuedCallContext : public BaseContext {
+  public:
+    EnqueuedCallContext(uint32_t context_id,
+                        AztecAddress address,
+                        AztecAddress msg_sender,
+                        std::span<const FF> calldata,
+                        bool is_static,
+                        std::unique_ptr<BytecodeManagerInterface> bytecode,
+                        std::unique_ptr<MemoryInterface> memory,
+                        EventEmitterInterface<ContextStackEvent>& ctx_stack_events)
+        : BaseContext(context_id,
+                      address,
+                      msg_sender,
+                      calldata,
+                      is_static,
+                      std::move(bytecode),
+                      std::move(memory),
+                      ctx_stack_events)
+    {}
+};
+
+// Parameters for a nested call need to be changed
+class NestedContext : public BaseContext {
+  public:
+    NestedContext(uint32_t context_id,
+                  AztecAddress address,
+                  AztecAddress msg_sender,
+                  std::span<const FF> calldata,
```

Comment:
So IIUC for the moment you split the classes but they are kind of the same and you'll be changing in other PR